### PR TITLE
C#: Fix `NOTIFICATION_PREDELETE` does not trigger `ExitTree`

### LINKED ISSUE #82279: C#: Script methods bound to `tree_exited` / `_ExitTree()` can't be found.
### Godot version

v4.2.dev5.mono.official [e3e2528ba]

### System information

Windows 11

### Issue description

After I upgraded my Godot version from 4.2dev3 to 4.2dev5 I encountered the issue that the `tree_exited` signal fires on nodes that have my custom C# script attached but it somehow already "forgot" that the Node actually had my script attached. So I get the error `emit_signalp: Error calling from signal 'tree_exited' to callable: 'Node2D::_on_tree_exited': Method not found.
  <C++ Source>   core/object/object.cpp:1125 @ emit_signalp()`

The same problem leads to the fact that my `public override void _ExitTree()` method in my script does not even run at all.

### Steps to reproduce

* Create a node with a C# script attached to it that implements either `_ExitTree()` or where the scene is connected to the `tree_exited` signal (calling a method on the same scene)
* Put debug prints into the signal-connected method or the `_ExitTree()` one
* Create a wrapper scene which contains the first scene
* Let the node of the first scene get removed by calling `QueueFree()` when clicking a button or pressing any key
* Get errors (in signal case) or no printing at all (in `_ExitTree()` case)

### Minimal reproduction project

Start the project and click the button. Notice that we don't get the log message from `_ExitTree()` and also an error from the `tree_exited` signal connection.
[tree-exit-bug.zip](https://github.com/godotengine/godot/files/12710513/tree-exit-bug.zip)

### PR DESCRIPTION
**Pull request by magian1127**
Fix #82279
It tests okay on Windows, and it tests using the program from #78634 and `Godot.exe --test` pass.

### CODE REVIEW

**Comment by akien-mga** [comment]
Despite the title this is a core change which can have significant implication for all languages, not just C#.

What's the problem with C# specifically that wouldn't affect GDScript or GDExtension? Or is there a more generalized problem?

**Comment by magian1127** [comment]
In C#, `QueueFree` doesn't trigger `ExitTree`,this is a serious bug for C #, but it does work in GDS.
I've been testing for a while, and so far, I haven't found this change affecting other languages. 
For the functionality in PR #78634, I have tested it with the provided test cases, and they all passed the tests.

**Comment by magian1127** [comment]
I tested again,  #78634 sorting in GDE is broken.

**Review by Sauermann — COMMENTED** [review]
This PR reverts parts of #78634 and when this PR gets merged in its current form, the following issues need to be reopened, because they will be reintroduced:
- https://github.com/godotengine/godot/issues/52325
- https://github.com/godotengine/godot-proposals/issues/5585

An alternative approach (and my suggestion) would be to investigate in which steps the attached C# script gets forgotten and make necessary adjustments there, so that the scripts are still remembered, when they get called.

**Comment by magian1127** [comment]
@Sauermann @akien-mga 
I changed the implementation method, only modifying C#. 
Now, the `reversed` postpones `CallDispose` until `NOTIFICATION_EXIT_TREE`. 
can now remove the `regression` tag.

**Comment by bitsawer** [comment]
Possibly related issues:
* https://github.com/godotengine/godot/issues/80834
* https://github.com/godotengine/godot/issues/6784

**Review by raulsntos — COMMENTED** [review]
This looks incorrect. Now it won't call `Dispose` when the notification call is reversed, at least until we receive notification `11` (which is `NOTIFICATION_EXIT_TREE` but we shouldn't be using an integer literal). However, this notification is only sent for Nodes, so that means Resources and RefCounteds won't be disposed now.

I don't think I understand issue https://github.com/godotengine/godot/issues/82279. Does it mean that `NOTIFICATION_PREDELETE` is being sent before `NOTIFICATION_EXIT_TREE`? That doesn't make sense to me, shouldn't the predelete notification be always the last one? The documentation says it can be used as a "destructor", so we should be able to dispose the C# object when receiving it.

Also, the issues mentioned by @bitsawer seem to affect GDScript as well, so maybe the issue is not exclusive to C# and then it should be fixed in Core for all languages.

**Comment by magian1127** [comment]
`csharp_script.cpp` `instance_create()`
``` c++
	return _create_instance(nullptr, 0, p_this, Object::cast_to<RefCounted>(p_this) != nullptr, unchecked_error);
```
Here the code shows that the `RefCounted` type sets `base_ref_counted` to `true`.
So in `CSharpInstance::notification`, the old code would not execute `CSharpInstanceBridge_CallDispose`.
Maybe I misunderstood the above?

File: modules/mono/csharp_script.cpp [hunk 1]
```diff
@@ -2003,6 +2003,15 @@ void CSharpInstance::notification(int p_notification, bool p_reversed) {
 
 		_call_notification(p_notification, p_reversed);
 
+		if (!p_reversed) {
+			GDMonoCache::managed_callbacks.CSharpInstanceBridge_CallDispose(
+					gchandle.get_intptr(), /* okIfNull */ false);
+		}
+
+		return;
+	} else if (predelete_notified && p_reversed) {
```
Comment (magian1127) [inline modules/mono/csharp_script.cpp:2012]: After testing, it is not necessary to evaluate `p_notification` here.

**Comment by raulsntos** [comment]
So I've looked a bit more into this, and this is what I've gathered.

`NOTIFICATION_PREDELETE` is sent when an Object is about to be deleted (called by `memdelete`). Then, Node overrides `_notification` and when it receives that notification it removes itself from its parent, thus sending `NOTIFICATION_EXIT_TREE`. Both notifications are sent with `p_reversed` set to true.

Since `NOTIFICATION_PREDELETE` is sent reversed, in a hierarchy like: `Object > Node > Node2D` the `_notification` method will be called first on Node2D, then Node. This means that `NOTIFICATION_EXIT_TREE` will be sent after the derived classes have received `NOTIFICATION_PREDELETE`. So in this case Node2D cleans up before receiving `NOTIFICATION_EXIT_TREE`.

I can imagine this being a problem. For example, if Node2D had some state that it clears when receiving `NOTIFICATION_PREDELETE`, then it won't be able to use it when receiving `NOTIFICATION_EXIT_TREE`. This looks backwards to me, I think `NOTIFICATION_PREDELETE` should always be the last one. Maybe we shouldn't send notifications inside `NOTIFICATION_PREDELETE`.

In the case of CSharpInstance, this means it will dispose the script before Node has a chance to send `NOTIFICATION_EXIT_TREE`, so by the time the notification is received there's no script to send it to. We want to make sure the C# script class is disposed **always** after all the classes in the hierarchy have received `NOTIFICATION_PREDELETE` and cleaned up (to ensure Node has sent `NOTIFICATION_EXIT_TREE` before disposing).

The current PR implementation is not good because it relies on receiving another notification after `NOTIFICATION_PREDELETE` which may work for Node because we know it will send `NOTIFICATION_EXIT_TREE` after, but I don't think we can rely on this behavior working for every class. Also, since it disposes the script in the _next_ notification, if there are other notifications only the first one will be received.

Maybe we could not dispose at all when receiving `NOTIFICATION_PREDELETE` and let the CSharpInstance destructor take care of it. But I think this is a risky change for 4.2, there may be consequences to this that I'm not able to see right now.

Alternatively, we could change `Object::notification` so that it calls `ScriptInstance::notification` always after `_notificationv`. This would mean for a hierarchy like `Object > Node > Node2D > ScriptA > ScriptB`, the notifications will be sent in the order `Node2D > Node > ScriptB > ScriptA`. Unfortunately, as mentioned earlier, this would reintroduce https://github.com/godotengine/godot/issues/52325.

We could also add a new notification that is sent by `memdelete` right after sending `NOTIFICATION_PREDELETE` and it would be used by CSharpInstance to dispose C# scripts, or by any other class that needs to ensure the cleanup happens after everything else. It feels weird because I think that's already the purpose of `NOTIFICATION_PREDELETE`, but the idea is that we wouldn't send any notifications inside the new one to ensure it really is the last notification. And this seems easier than refactoring all the existing code that uses `NOTIFICATION_PREDELETE` to not send notifications (like Node does now with `NOTIFICATION_EXIT_TREE`).

**Comment by magian1127** [comment]
The inability to trigger "ExitTree" has already seriously affected the continued use of Godot. If the current PR's solution is not suitable, hope that others can provide a better one.

**Pull request closed without merging**